### PR TITLE
chore: Scaffold meta-transaction v2 endpoints [RFQ-768]

### PR DESCRIPTION
**Pull request by Haozhuo**
# Description

Scaffold meta-transaction v2 endpoints in meta-transaction router, handler and service. Most of the changes are a copy of existing meta-transaction endpoints & functions with minor modifications. They will be further modified in the coming weeks but for now, scaffolding is sufficient. 

The new changes are mainly adding fees to request & response types and parsing fees.

# Testing & Simbot Run

# Checklist

-   [x] Update 0x API documentation (gitbook) if needed (e.g. public facing API change).
-   [x] Add tests to cover changes as needed.
-   [x] All dependent changes (e.g. RFQ server, Gas Price Oracle) have been deployed (if any).

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -256,11 +256,28 @@ export interface MetaTransactionV1QuoteResponse extends BasePriceResponse {
     orders?: any;
 }
 
+/**
+ * Response type for /meta_transaction/v2/quote endpoint
+ */
+export interface MetaTransactionV2QuoteResponse extends BasePriceResponse {
```
Comment (patrick-dowell) [inline src/types.ts:262]: This may be a dumb question, but what's the difference between quote and price?

File: src/types.ts [hunk 2]
```diff
@@ -289,6 +313,13 @@ export interface MetaTransactionV1QuoteParams extends SwapQuoteParamsBase {
     takerAddress: string;
 }
 
+/**
+ * Parameters for the V2 Meta Transaction Service price and quote functions.
+ */
+export interface MetaTransactionV2QuoteParams extends MetaTransactionV1QuoteParams {
```
Comment (patrick-dowell) [inline src/types.ts:319]: why is there a MetaTransactionV2QuoteParams and MetaTransactionsV2QuoteRequestParams? Seems like you should only need one or the other.

File: src/services/meta_transaction_service.ts [hunk 3]
```diff
@@ -109,6 +166,79 @@ export class MetaTransactionService implements IMetaTransactionService {
...
+            shouldSellEntireBalance: false,
+            skipValidation: true,
+        };
+
+        const quote = await this._swapService.calculateSwapQuoteAsync(quoteParams);
+
+        // Quote Report
+        if (endpoint === 'quote' && quote.extendedQuoteReportSources && kafkaProducer) {
+            const quoteId = getQuoteIdFromSwapQuote(quote);
+            publishQuoteReport(
```
Comment (patrick-dowell) [inline src/services/meta_transaction_service.ts:196]: should we publish fee info to kafka in the quote report?

File: src/handlers/meta_transaction_handlers.ts [hunk 4]
```diff
@@ -237,3 +373,129 @@ function parseV1RequestParams(req: express.Request): MetaTransactionV1QuoteReque
...
+
+    if (req.query.feeConfigs) {
+        const feeConfigs = req.query.feeConfigs as ParsedQs;
+        parsedFeeConfigs = {};
+
+        // Parse the integrator fee config
+        if (feeConfigs.integrator) {
+            const integratorFee: ParsedQs = feeConfigs.integrator as ParsedQs;
+
+            if (integratorFee.kind !== 'volume') {
```
Comment (patrick-dowell) [inline src/handlers/meta_transaction_handlers.ts:405]: isn't volume the only kind for integratorFee?

File: src/handlers/meta_transaction_handlers.ts [hunk 5]
```diff
@@ -237,3 +373,129 @@ function parseV1RequestParams(req: express.Request): MetaTransactionV1QuoteReque
...
+                    {
+                        field: 'feeConfigs',
+                        code: ValidationErrorCodes.IncorrectFormat,
+                        reason: ValidationErrorReasons.InvalidGaslessFeeKind,
+                    },
+                ]);
+            }
+
+            const feePercentage = new BigNumber(zeroexFee.volumePercentage as string);
+            if (feePercentage.gt(1)) {
```
Comment (patrick-dowell) [inline src/handlers/meta_transaction_handlers.ts:451]: Do 100% fees even make sense?

File: src/handlers/meta_transaction_handlers.ts [hunk 6]
```diff
@@ -237,3 +373,129 @@ function parseV1RequestParams(req: express.Request): MetaTransactionV1QuoteReque
...
+                    integratorSharePercentage: new BigNumber(zeroexFee.integratorSharePercentage as string),
+                };
+            }
+        }
+
+        // Parse the gas fee config
+        if (feeConfigs.gas) {
+            const gasFee: ParsedQs = feeConfigs.gas as ParsedQs;
+
+            if (gasFee.kind !== 'gas') {
```
Comment (patrick-dowell) [inline src/handlers/meta_transaction_handlers.ts:482]: can gasFee ever not be gas?

File: src/handlers/meta_transaction_handlers.ts [hunk 7]
```diff
@@ -39,6 +47,134 @@ export class MetaTransactionHandlers {
         this._metaTransactionService = metaTransactionService;
     }
 
+    /**
+     * Handler for the /meta_transaction/v2/quote endpoint
+     */
+    public async getV2QuoteAsync(req: express.Request, res: express.Response): Promise<void> {
```
Comment (patrick-dowell) [inline src/handlers/meta_transaction_handlers.ts:53]: how much shared code is there between this logic and the v1QuoteAsync logic? Same for v1PriceAsync?

**Review by patrick-dowell — COMMENTED** [review]
Mostly just comments. The code itself looks reasonable to me, but I do have questions!

File: src/types.ts (hunk 1 above)
Comment (Haozhuo) [reply src/types.ts:262]: No it's a great question! In general, `/quote` endpoint is intended to be called when there is an intention to trade (there are more processing we do on our end). If not, `/price` endpoint should be called if you're just loosely checking. This is a pattern we use in our other APIs as well.

In the case of meta-transaction, `/price` and `/quote` are similar but no meta-transaction object is returned & no quote report is generated in `/price` (there are some other minor differences but not huge).

File: src/types.ts (hunk 2 above)
Comment (Haozhuo) [reply src/types.ts:319]: `MetaTransactionsV2QuoteRequestParams` is the type after parsing the incoming request to `/quote` (same for `MetaTransactionsV1QuoteRequestParams`). 
`MetaTransactionV2QuoteParams` is the type being passed to  `getMetaTransactionV2QuoteAsync` of `MetaTransactionService` (same for `MetaTransactionV1QuoteParams`).

The intention here is to keep it consistent with V1 types.

I think the intention on why V1 is structured this way is to have a type that purely represents the incoming request (`QuoteRequestParams`). 
Once we have the parsed request, we can do some processing and fields to it and thus another type (`QuoteParams`). The new type represents a process / modified object from the incoming request which is not changed, thus the distinction of the 2 types.

File: src/services/meta_transaction_service.ts (hunk 3 above)
Comment (Haozhuo) [reply src/services/meta_transaction_service.ts:196]: Yep. We should add it after the fee calculation & fee storage code are completed. For scaffolding we can't add it now as the necessary code has been committed yet

File: src/handlers/meta_transaction_handlers.ts (hunk 4 above)
Comment (Haozhuo) [reply src/handlers/meta_transaction_handlers.ts:405]: Yes for now. Here trying to add an extra validation here for the case that the caller might provide the wrong `kind` and provide error code.

File: src/handlers/meta_transaction_handlers.ts (hunk 5 above)
Comment (Haozhuo) [reply src/handlers/meta_transaction_handlers.ts:451]: Good catch! Will change it to `gte`

File: src/handlers/meta_transaction_handlers.ts (hunk 6 above)
Comment (Haozhuo) [reply src/handlers/meta_transaction_handlers.ts:482]: If no mistake is made, then `gasFee` should always have kind `gas`. Again trying to be extra safe here in the case of mistake and surface the error to the caller

File: src/handlers/meta_transaction_handlers.ts (hunk 7 above)
Comment (Haozhuo) [reply src/handlers/meta_transaction_handlers.ts:53]: Assuming you're talking about handler level, parsing of incoming request and get meta-transaction v2 quote are different. The majority of the shared code on handler level is error handling

**Comment by Haozhuo** [comment]
https://linear.app/0xproject/issue/RFQ-768/meta-transaction-v2-scaffolding

**Review by phil-ociraptor — APPROVED** [review]
Approving the scaffolding for v2 - for the service layer, let's try to keep v1 and v2 unified where we can

File: src/handlers/meta_transaction_handlers.ts (hunk 7 above)
Comment (Haozhuo) [reply src/handlers/meta_transaction_handlers.ts:53]: Got it. I will do another refactoring after merging this scaffolding PR. Want to limit the scope of the scaffolding

**Pull request merged**